From 51bf645d72e99516a0c7c51d3efb679c3235a211 Mon Sep 17 00:00:00 2001
From: Zhang Qiao <zhangqiao22@huawei.com>
Date: Thu, 26 May 2022 20:36:27 +0800
Subject: [PATCH] sched/fair: Update rq clock before unthrottle a qos cfs_rq
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 51bf645d72e99516a0c7c51d3efb679c3235a211
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I4VZJT
CVE: NA

--------------------------------

------------[ cut here]------------
 rq->clock_update_flags < RQCF_ACT_SKIP
 WARNING: CPU: 5 PID: 3312 at kernel/sched/sched.h:1223 update_curr+0x1e5/0x210
 CPU: 5 PID: 3312 Comm: a.out Tainted: G S5.10.0.zq+ #1
 Hardware name: Huawei RH2288H V3/BC11HGSA0, BIOS 3.35 10/20/2016
 RIP: 0010:update_curr+0x1e5/0x210
 enqueue_entity+0x378/0xd00
 unthrottle_qos_cfs_rq+0x1bc/0x2a0
 __unthrottle_qos_cfs_rqs+0x87/0xa0
 qos_overload_timer_handler+0x35/0x60
 __run_hrtimer+0x5e/0x190
 __hrtimer_run_queues+0x81/0xe0
 hrtimer_interrupt+0x110/0x2c0
 __sysvec_apic_timer_interrupt+0x5f/0xd0
 sysvec_apic_timer_interrupt+0x31/0x80
 asm_sysvec_apic_timer_interrupt+0x12/0x20

After the last rq_pin_lock(), there is no rq clock update before calling
enqueue_entity() at unthrottle_qos_cfs_rq();

This patch fixes it by updating rq clock before calling enqueue_entity().

Fixes: c62a5f1384b9("sched/qos: Add qos_tg_{throttle,unthrottle}_{up,down}")
Signed-off-by: Zhang Qiao <zhangqiao22@huawei.com>
Reviewed-by: Chen Hui <judy.chenhui@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 kernel/sched/fair.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index 5fe13efce378..9d5c780160c5 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -7266,6 +7266,7 @@ static void unthrottle_qos_cfs_rq(struct cfs_rq *cfs_rq)
 
 	cfs_rq->throttled = 0;
 
+	update_rq_clock(rq);
 	list_del_init(&cfs_rq->qos_throttled_list);
 
 	/* update hierarchical throttle state */
-- 
2.33.0

